### PR TITLE
fix(api): keep at least one success run in purge

### PR DESCRIPTION
**Pull request by bnjjj**
Signed-off-by: Benjamin Coenen <benjamin.coenen@corp.ovh.com>

1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/api/workflow/dao_run_test.go [hunk 1]
```diff
@@ -135,6 +136,175 @@ func TestPurgeWorkflowRun(t *testing.T) {
...
+	proj, _ = project.LoadByID(db, cache, proj.ID, u, project.LoadOptions.WithApplications, project.LoadOptions.WithPipelines, project.LoadOptions.WithEnvironments, project.LoadOptions.WithGroups)
+
+	w := sdk.Workflow{
+		Name:       "test_purge_1",
+		ProjectID:  proj.ID,
+		ProjectKey: proj.Key,
+		Root: &sdk.WorkflowNode{
+			Pipeline: pip,
+			Triggers: []sdk.WorkflowNodeTrigger{
+				sdk.WorkflowNodeTrigger{
```
Comment (golangci) [inline engine/api/workflow/dao_run_test.go]: File is not gofmt-ed with -s

File: engine/api/workflow/dao_run.go [hunk 2]
```diff
@@ -553,10 +579,52 @@ func PurgeWorkflowRun(db gorp.SqlExecutor, wf sdk.Workflow) error {
...
+	successIDs := []struct {
+		ID string `db:"id"`
+	}{}
+	if _, errS := db.Select(&successIDs, querySuccessIds, wf.ID, strings.Join(filteredPurgeTags, ","), sdk.StatusSuccess.String()); errS != nil {
+		log.Warning("PurgeWorkflowRun> Unable to get workflow run in success for purge with workflow id %d, tags %v and history length %d : %s", wf.ID, wf.PurgeTags, wf.HistoryLength, errS)
+		return errS
+	}
+
 	idsToUpdate := []string{}
+	fmt.Printf("%+v\n", successIDs)
```
Comment (sguiheux) [inline engine/api/workflow/dao_run.go]: remove

File: engine/api/workflow/dao_run.go [hunk 3]
```diff
@@ -525,26 +525,52 @@ func PurgeWorkflowRun(db gorp.SqlExecutor, wf sdk.Workflow) error {
...
+
 		qDelete := `
 			UPDATE workflow_run SET to_delete = true
 			WHERE workflow_run.id IN (
-				SELECT workflow_run.id FROM workflow_run
+				SELECT workflow_run.id
+					FROM workflow_run
 				WHERE workflow_run.workflow_id = $1
-				ORDER BY workflow_run.id DESC OFFSET $2 ROWS
+				AND workflow_run.id <> $2
```
Comment (yesnault) [inline engine/api/workflow/dao_run.go]: "<"

File: engine/api/workflow/dao_node_run.go [hunk 4]
```diff
@@ -465,6 +465,7 @@ func GetNodeRunBuildCommits(db gorp.SqlExecutor, store cache.Store, p *sdk.Proje
 		log.Info("GetNodeRunBuildCommits> Current Build number: %d - Current Hash: %s - Previous Build number: %d - Previous Hash: %s", cur.BuildNumber, cur.Hash, prev.BuildNumber, prev.Hash)
 	}
 
+	log.Debug("GetNodeRunBuildCommits> current : %+v , previous: %+v", cur, prev)
```
[the remark this answers is not in the document]
Comment (bnjjj) [reply engine/api/workflow/dao_node_run.go]: TODO added, I need this one if the bug is not fixed, I will delete it after for sure

**Pull request merged**